### ISSUE TITLE
Bug: can't route when query string contains latin accents

### ISSUE THREAD

**Issue by xfra35** [bcosca/fatfree-core#126]
Bug reported by @WLR86. Test case: https://github.com/WLR86/F3-reproducing-bug

In short: when calling `/path?foo=barré`, the framework cannot route the request if `é` is not encoded in UTF-8. The case may arise if the query string comes from a non UTF-8 form.

The bug comes from the use of the regex `u` modifier in [mask()](https://github.com/bcosca/fatfree-core/blob/b58004f6f4489d77beeda72f01cbfab167cffa83/base.php#L1388).

According to the [docs](http://php.net/manual/en/reference.pcre.pattern.modifiers.php):

> UTF-8 validity of the pattern and the subject is checked since PHP 4.3.5. An invalid subject will cause the preg_* function to match nothing.

In other words, the `u` modifier can't be used on a non UTF-8 URL.

I suggest to remove that modifier when `ENCODING` is different from `UTF-8`.

**Comment by ikkez** [comment]
The query string should not be important for the mask function. Couldn't we just use `$url=$this->hive['PATH'];`instead? Or is the URI path affected as well?

**Comment by xfra35** [comment]
I don't know if we can assume that `PATH` will always be UTF-8-encoded.

I'm also wondering why `mask()` is working on the query string. It seems that it's there to feed `PARAMS.0`, but I don't really see the use of it.

**Comment by kavin-90** [comment]
I don't see any Problem with that /barré

Tested works 100% with 3.5.1 with my fix

**Comment by WLR86** [comment]
@kavin-90 Are you using the files from the test case ?

It's not just about typing "/barré" in your url bar, it's about using a different encoding than UTF-8 and attempting to use the provide form with "barré" as a value, which is quite not the same thing.

The generated URL would then be "dummy?searchQuery=barr%E9" (note that "é" is being url_encoded ) and this is what you got to test.

So please test this very same case and tell us what you get

**Comment by kavin-90** [comment]
Yes, Method POST with F3 using UTF8 char set.

**Comment by WLR86** [comment]
Ok @kavin-90 , I know it works using POST and UTF-8 - this case deals with GET and ISO which means NO POST and NO UTF-8

@xfra35 by the way replacing the url_encoded value by its raw value don't break the routing on my side : `?foo=barré `would just return a UTF-8 encoded value for foo : 'barrÃ©' 
So yes the routing works in this example, but still, considering the encoding setting, value shouldn't be returned as an UTF-8 encoded string.

Again, on my side routing breaks on this kind of querystring : "dummy?searchQuery=barr%E9"

**Comment by kavin-90** [comment]
Tested and works.

```
$.post( "http://localhost/m/public/barré",{data : 'barré'},function( data ) {
if(data.msg) alert(data.msg);
});
```

Message retun
Encoding is ISO-8859-15 Message is barré

**Comment by WLR86** [comment]
@kavin-90 yup confirmed this works on my side too
Please test this : 
`?searchQuery=barr%E9`

**Comment by kavin-90** [comment]
`Encoding is ISO-8859-15 Message is barré and query is barr%E9`

**Comment by WLR86** [comment]
Please @kavin-90 your code uses POST method instead of GET

Again, it works just fine using POST, the problem occurs when using GET

**Comment by kavin-90** [comment]
Browser default charset is UTF8

**Comment by kavin-90** [comment]
Checked works with GET

**Comment by WLR86** [comment]
You mean you actually tested the provided code and it worked ?
I updated the repo with code that allows to test both methods

**Comment by kavin-90** [comment]
Yes your repo works fine.

**Comment by WLR86** [comment]
Ok thanks

**Comment by xfra35** [comment]
Guys, here's a simpler code to reproduce the issue:

```php
$f3=require('lib/base.php');

$f3->route('GET /',function($f3){
    echo 'QUERY = '.$f3->QUERY;
});

$f3->run();
```

Responses:
* `/` => 200
* `/?foo=barre` => 200
* `/?foo=barr%C3%A9` => 200
* `/?foo=barr%E9` => 404

NB: `é` is encoded as `%C3%A9` in UTF-8 and `%E9` in ISO-8859-15.

**Comment by xfra35** [comment]
Same issue with:

```php
$f3->route('GET /path/*',function($f3,$params){
    echo 'PATH = '.$f3->PATH;
});
```

Responses:
* `/path/barre` => 200
* `/path/barr%C3%A9` => 200
* `/path/barr%E9` => 404

**BUT** according to [RFC 3986](http://tools.ietf.org/html/rfc3986#section-2.5):
> the data should first be encoded as octets according to the UTF-8 character encoding [STD63]; then only those octets that do not correspond to characters in the unreserved set should be percent-encoded

So if I understand it well, `/path/barr%E9` is an invalid URI. In which case, the framework behaviour is fine.

**Comment by xfra35** [comment]
Should be fixed now.

**Comment by kavin-90** [comment]
@xfra35 checked i got 200 for all

Responses:
```

    / => 200
    /?foo=barre => 200
    /?foo=barr%C3%A9 => 200
    /?foo=barr%E9 => 200
```
